Sources/NIOWebSocket/NIOWebSocketServerUpgrader.swift

```diff
@@ -224,62 +224,177 @@ public final class NIOWebSocketServerUpgrader: HTTPServerProtocolUpgrader, @unch
+    public init(
+        maxFrameSize: Int = 1 << 14,
+        automaticErrorHandling: Bool = true,
+        shouldUpgrade: @escaping @Sendable (Channel, HTTPRequestHead) -> EventLoopFuture<HTTPHeaders?>,
+        upgradePipelineHandler: @escaping @Sendable (Channel, HTTPRequestHead) -> EventLoopFuture<UpgradeResult>
+    ) {
+        precondition(maxFrameSize <= UInt32.max, "invalid overlarge max frame size")
+        self.shouldUpgrade = shouldUpgrade
+        self.upgradePipelineHandler = upgradePipelineHandler
+        self.maxFrameSize = maxFrameSize
+        self.automaticErrorHandling = automaticErrorHandling
     }
 
-    public func upgrade(context: ChannelHandlerContext, upgradeRequest: HTTPRequestHead) -> EventLoopFuture<Void> {
-        /// We never use the automatic error handling feature of the WebSocketFrameDecoder: we always use the separate channel
-        /// handler.
-        var upgradeFuture = context.pipeline.addHandler(WebSocketFrameEncoder()).flatMap {
-            context.pipeline.addHandler(ByteToMessageHandler(WebSocketFrameDecoder(maxFrameSize: self.maxFrameSize)))
+    public func buildUpgradeResponse(
+        channel: Channel,
+        upgradeRequest: HTTPRequestHead,
+        initialResponseHeaders: HTTPHeaders
+    ) -> EventLoopFuture<HTTPHeaders> {
+        _buildUpgradeResponse(
+            channel: channel,
+            upgradeRequest: upgradeRequest,
+            initialResponseHeaders: initialResponseHeaders,
+            shouldUpgrade: self.shouldUpgrade
+        )
+    }
+
+    public func upgrade(channel: Channel, upgradeRequest: HTTPRequestHead) -> EventLoopFuture<UpgradeResult> {
+        _upgrade(
+            channel: channel,
+            upgradeRequest: upgradeRequest,
+            maxFrameSize: self.maxFrameSize,
+            automaticErrorHandling: self.automaticErrorHandling,
+            upgradePipelineHandler: self.upgradePipelineHandler
+        )
+    }
+}
+
+private func _buildUpgradeResponse(
+    channel: Channel,
+    upgradeRequest: HTTPRequestHead,
+    initialResponseHeaders: HTTPHeaders,
+    shouldUpgrade: @Sendable (Channel, HTTPRequestHead) -> EventLoopFuture<HTTPHeaders?>
+) -> EventLoopFuture<HTTPHeaders> {
+    let key: String
+    let version: String
+
+    do {
+        key = try upgradeRequest.headers.nonListHeader("Sec-WebSocket-Key")
+        version = try upgradeRequest.headers.nonListHeader("Sec-WebSocket-Version")
+    } catch {
+        return channel.eventLoop.makeFailedFuture(error)
+    }
+
+    // The version must be 13.
+    guard version == "13" else {
+        return channel.eventLoop.makeFailedFuture(NIOWebSocketUpgradeError.invalidUpgradeHeader)
+    }
+
+    return shouldUpgrade(channel, upgradeRequest).flatMapThrowing { extraHeaders in
+        guard let extraHeaders = extraHeaders else {
+            throw NIOWebSocketUpgradeError.unsupportedWebSocketTarget
         }
+        return extraHeaders
+    }.map { (extraHeaders: HTTPHeaders) in
```

Comment:
This map seems redundant; can't we just do all of this work in the `flatMapThrowing`?